### PR TITLE
jgrss/batch val metric logging

### PR DESCRIPTION
**Pull request by jgrss**
This PR add logging for batch validation metrics.

### CODE REVIEW

File: src/cultionet/callbacks.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
+import typing as T
```
Comment (jgrss) [inline src/cultionet/callbacks.py:1]: Ignore `callbacks` because this was simply moved from another module.

File: src/cultionet/models/lightning.py [hunk 2]
```diff
@@ -909,10 +912,40 @@ def validation_step(self, batch: Data, batch_idx: int = None) -> dict:
         }
         if 'crop_type_f1' in eval_metrics:
             metrics['vctf1'] = eval_metrics['crop_type_f1']
+
         self.log_dict(metrics, on_step=False, on_epoch=True, prog_bar=True)
 
+        if self.save_batch_val_metrics:
```
Comment (jgrss) [inline src/cultionet/models/lightning.py:918]: New call in validation step

File: src/cultionet/models/lightning.py [hunk 3]
```diff
@@ -909,10 +912,40 @@ def validation_step(self, batch: Data, batch_idx: int = None) -> dict:
...
+            self._save_batch_metrics(metrics, self.current_epoch, batch)
+
         return metrics
 
+    def _save_batch_metrics(
```
Comment (jgrss) [inline src/cultionet/models/lightning.py:923]: New function to save batch validation metrics to file.

File: src/cultionet/scripts/args.yml [hunk 4]
```diff
@@ -584,6 +584,12 @@ train:
     kwargs:
       default: !!null
       type: '&int'
+  save_batch_val_metrics:
```
Comment (jgrss) [inline src/cultionet/scripts/args.yml:587]: New CLI arg

File: src/cultionet/model.py [hunk 5]
```diff
@@ -187,7 +189,7 @@ def fit_maskrcnn(
         devices=1 if device == 'gpu' else None,
         num_processes=0,
         accelerator=device,
-        log_every_n_steps=10,
+        log_every_n_steps=50,
```
Comment (nnguyen622) [inline src/cultionet/model.py:192]: oh we increase this number quite a lot

File: src/cultionet/models/lightning.py [hunk 6]
```diff
@@ -909,10 +912,40 @@ def validation_step(self, batch: Data, batch_idx: int = None) -> dict:
...
+        self,
+        metrics: T.Dict[str, torch.Tensor],
+        epoch: int,
+        batch: Data
+    ) -> None:
+        """Saves batch metrics to a parquet file
+        """
+        if not self.trainer.sanity_checking:
+            write_metrics = {
+                'epoch': [epoch] * len(batch.train_id),
```
Comment (nnguyen622) [inline src/cultionet/models/lightning.py:933]: interesting, so the epoch here is multiple by the length of all the training data, including the augmentation?

**Review by nnguyen622 — APPROVED** [review]
Thanks @jgrss , this looks like a great addition to check model version.
I leave some minor comments but these changes look good👍

File: src/cultionet/model.py (hunk 5 above)
Comment (jgrss) [reply src/cultionet/model.py:192]: Yeah, it simply reduces the number of hits to the logger. The same metrics are tracked.

File: src/cultionet/models/lightning.py (hunk 6 above)
Comment (jgrss) [reply src/cultionet/models/lightning.py:933]: Not all of the training data, just the training data in the batch. E.g., each batch (if batch size = 8) will have 8 different training ids, which could be any of the data--raw or augmented.

Because metrics are aggregated for each batch, each data id in a batch will have the same metric as all data ids in the batch. However, data sampling is random, so a data id will not likely have the same combination of batch ids each epoch. And because data batches are grouped, this is about as detailed as we can get without becoming inefficient.

Comment (nnguyen622) [reply src/cultionet/models/lightning.py:933]: Thanks for the explanation Jordan 👍 
The log with each batch id is very detailed 👍
